File: C++/Heap_PriorityQueue_Medium_TaskScheduler_621.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) 
    {
        vector<int> counts(26, 0);
        for(int i = 0; i < tasks.size(); i++)
        {
            counts[tasks[i] - 'A']++;
        }

        priority_queue<int> pq;
        for(int count : counts)
        {
            if(count > 0)
            {
                pq.push(count);
            }
        }

        int total_time = 0;

        while(!pq.empty())
        {
            vector<int> waitlist;
            int cycle_size = n + 1;
            int tasks_done_this_cycle = 0;

            while(cycle_size > 0 && !pq.empty())
            {
                int current_task_count = pq.top();
                pq.pop();

                current_task_count--;
                waitlist.push_back(current_task_count);

                tasks_done_this_cycle++;
                cycle_size--;
            }

            for(int remaining_count : waitlist)
            {
                if(remaining_count > 0)
                {
                    pq.push(remaining_count);
                }
            }

            if(pq.empty())
            {
                total_time += tasks_done_this_cycle;
            }
            else
            {
                total_time += (n + 1);
            }
        }

        return total_time;
    }
};

File: Python/Heap_PriorityQueue_Medium_TaskScheduler_621.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        count = [0] * 26

        for i in range(0, len(tasks)):
            count[ord(tasks[i]) - ord('A')] += 1

        heap = []
        for num in count:
            if num > 0:
                heapq.heappush(heap, -num)

        totalTime = 0

        while heap:
            waiting_line = []
            time_in_that_window = 0
            tasks_to_fill = n + 1

            while tasks_to_fill > 0 and heap:
                waiting_line.append(-heapq.heappop(heap) - 1)
                tasks_to_fill -= 1
                time_in_that_window += 1

            for remaining in waiting_line:
                if remaining > 0:
                    heapq.heappush(heap, -remaining)
            
            if not heap:
                totalTime += time_in_that_window
            else:
                totalTime += (n + 1)

        return totalTime




